### DISCUSSION TITLE
curl: (60) schannel: CertGetCertificateChain trust error CERT_TRUST_IS_UNTRUSTED_ROOT

### DISCUSSION THREAD

**Discussion by ptcalex** [curl/curl#15328] [General]
I have created a self-signed CA certificate `ca.cer` and an HTTPS server certificate signed by it.
I'm using this curl version on Windows 11:

```
curl 8.8.0 (Windows) libcurl/8.8.0 Schannel zlib/1.3 WinIDN
Release-Date: 2024-05-22
Protocols: dict file ftp ftps http https imap imaps ipfs ipns mqtt pop3 pop3s smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS HSTS HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM SPNEGO SSL SSPI threadsafe Unicode UnixSockets
```

I run 3 commands, and I'm expecting only the first one to fail.

**First**: this one fails because `ca.cer` is not known to the curl environment

```
C:\> curl --ssl-no-revoke -X POST https://%servername%:%port%
curl: (60) schannel: CertGetCertificateChain trust error CERT_TRUST_IS_UNTRUSTED_ROOT
```

**Second**: this one succeeds because I pass `ca.cer` to `--cacert`:
```
C:\>curl --cacert ca.cer --ssl-no-revoke -X POST https://%servername%:%port%
```

**Third**: I then install `ca.cer` into the Windows Certificates - Local Computer / Trusted Root Certification Authorities / Certificates, and run this third command, which fails:

```
C:\>curl --ca-native --ssl-no-revoke -X POST https://%servername%:%port%
curl: (60) schannel: CertGetCertificateChain trust error CERT_TRUST_IS_UNTRUSTED_ROOT
```

Why does the third command fails? I can see `ca.cer` among the trusted roots, and have used `--ca-native`.

At https://curl.se/docs/sslcerts.html I read that _If curl was built with Schannel, Secure Transport or were instructed to use the native CA Store, then curl uses the certificates that are built into the OS_.

**Comment by jay** [comment]
> Why does the third command fails? I can see `ca.cer` among the trusted roots, and have used `--ca-native`.

[--ca-native](https://curl.se/docs/manpage.html#--ca-native) is not needed: "When curl on Windows is built to use Schannel, this feature is implied and curl then only uses the native CA store."

Try `certutil -user -verify <servercert>` as the same user that runs the curl command and look for errors other than revocation like "A certificate chain could not be built to a trusted root authority."

**Reply by ptcalex** [reply]
Here's the output of the command:

```
C:\>certutil -user -verify ca.cer
Issuer:
    CN=XXX
    C=XX
  Name Hash(sha1): xxx
  Name Hash(md5): xxx
Subject:
    CN=XXX
    C=XX
  Name Hash(sha1): xxx
  Name Hash(md5): xxx
Cert Serial Number: xxx

dwFlags = CA_VERIFY_FLAGS_CONSOLE_TRACE (0x20000000)
dwFlags = CA_VERIFY_FLAGS_DUMP_CHAIN (0x40000000)
ChainFlags = CERT_CHAIN_REVOCATION_CHECK_CHAIN_EXCLUDE_ROOT (0x40000000)
HCCE_CURRENT_USER
CERT_CHAIN_POLICY_BASE
-------- CERT_CHAIN_CONTEXT --------
ChainContext.dwInfoStatus = CERT_TRUST_HAS_PREFERRED_ISSUER (0x100)

SimpleChain.dwInfoStatus = CERT_TRUST_HAS_PREFERRED_ISSUER (0x100)

CertContext[0][0]: dwInfoStatus=10c dwErrorStatus=0
  Issuer: CN=XXX, C=XX
  NotBefore: 17/10/2024 11:01
  NotAfter: 17/10/2044 11:01
  Subject: CN=XXX, C=XX
  Serial: XXX
  Cert: XXX
  Element.dwInfoStatus = CERT_TRUST_HAS_NAME_MATCH_ISSUER (0x4)
  Element.dwInfoStatus = CERT_TRUST_IS_SELF_SIGNED (0x8)
  Element.dwInfoStatus = CERT_TRUST_HAS_PREFERRED_ISSUER (0x100)

Exclude leaf cert:
  Chain: XXX
Full chain:
  Chain: XXX
------------------------------------
Verified Issuance Policies: All
Verified Application Policies: All
Cert is a CA certificate
Cannot check leaf certificate revocation status
CertUtil: -verify command completed successfully.

```

**Reply by jay** [reply]
Check the server certificate not the CA certificate

**Reply by ptcalex** [reply]
When I check the server certificate I get this:

```
Issuer:
    CN=XXX
    C=XX
  Name Hash(sha1): c33xxx
  Name Hash(md5): 95dxxx
Subject:
    CN=XXX
    C=XX
  Name Hash(sha1): d3axxx
  Name Hash(md5): 4daxxx
Cert Serial Number: 671xxx

dwFlags = CA_VERIFY_FLAGS_CONSOLE_TRACE (0x20000000)
dwFlags = CA_VERIFY_FLAGS_DUMP_CHAIN (0x40000000)
ChainFlags = CERT_CHAIN_REVOCATION_CHECK_CHAIN_EXCLUDE_ROOT (0x40000000)
HCCE_CURRENT_USER
CERT_CHAIN_POLICY_BASE
-------- CERT_CHAIN_CONTEXT --------
ChainContext.dwInfoStatus = CERT_TRUST_HAS_PREFERRED_ISSUER (0x100)
ChainContext.dwErrorStatus = CERT_TRUST_REVOCATION_STATUS_UNKNOWN (0x40)

SimpleChain.dwInfoStatus = CERT_TRUST_HAS_PREFERRED_ISSUER (0x100)
SimpleChain.dwErrorStatus = CERT_TRUST_REVOCATION_STATUS_UNKNOWN (0x40)

CertContext[0][0]: dwInfoStatus=104 dwErrorStatus=40
  Issuer: CN=XXX, C=XX
  NotBefore: 17/10/2024 11:06
  NotAfter: 17/10/2044 11:06
  Subject: CN=XXX, C=XX
  Serial: 671xxx
  SubjectAltName: DNS Name=XXX
  Cert: 7abxxx
  Element.dwInfoStatus = CERT_TRUST_HAS_NAME_MATCH_ISSUER (0x4)
  Element.dwInfoStatus = CERT_TRUST_HAS_PREFERRED_ISSUER (0x100)
  Element.dwErrorStatus = CERT_TRUST_REVOCATION_STATUS_UNKNOWN (0x40)

CertContext[0][1]: dwInfoStatus=10c dwErrorStatus=0
  Issuer: CN=XXX, C=XX
  NotBefore: 17/10/2024 11:01
  NotAfter: 17/10/2044 11:01
  Subject: CN=XXX, C=XX
  Serial: 671xxx
  Cert: 127xxx
  Element.dwInfoStatus = CERT_TRUST_HAS_NAME_MATCH_ISSUER (0x4)
  Element.dwInfoStatus = CERT_TRUST_IS_SELF_SIGNED (0x8)
  Element.dwInfoStatus = CERT_TRUST_HAS_PREFERRED_ISSUER (0x100)

Exclude leaf cert:
  Chain: 7abxxx
Full chain:
  Chain: 50cxxx
  Issuer: CN=XXX, C=XX
  NotBefore: 17/10/2024 11:06
  NotAfter: 17/10/2044 11:06
  Subject: CN=XXX, C=XX
  Serial: 671xxx
  SubjectAltName: DNS Name=XXX
  Cert: 7abxxx
The revocation function was unable to check revocation for the certificate. 0x80092012 (-2146885614 CRYPT_E_NO_REVOCATION_CHECK)
------------------------------------
Revocation check skipped -- no revocation information available
Cannot check leaf certificate revocation status
CertUtil: -verify command completed successfully.

```

**Reply by jay** [reply]
Ok that looks normal except that it's missing the server authentication OID which might be why verification fails. Check your server certificate for that OID

![Capture](https://github.com/user-attachments/assets/9fd2706b-f7e6-4915-920e-72830c62f2a3)

**Reply by jay** [reply]
> C:\>curl --ca-native --ssl-no-revoke -X POST https://%servername%:%port%
> curl: (60) schannel: CertGetCertificateChain trust error CERT_TRUST_IS_UNTRUSTED_ROOT

Looking into this further, that error comes from Curl_verify_certificate which libcurl uses for manual certificate verification only if the OS is Windows CE or the user has chosen manual verification. That means you should not see that error from that command, unless maybe there is something in your curlrc that causes it.

https://github.com/curl/curl/blob/7eb8c048470ed2cc14dca75be9c1cdae7ac8498b/lib/vtls/schannel.c#L1105-L1118

I suggest use [--disable](https://curl.se/docs/manpage.html#-q) as the first parameter to disable curlrc and also use --verbose

~~~
curl --disable --verbose --ssl-no-revoke -X POST https://%servername%:%port%
~~~

**Reply by ptcalex** [reply]
Yes, thank you. When I add `--disable` I can verify it successfully.
This is the curl version shipped with Windows. I noticed that if I use another curl version I could validate.
So that means that the curl shipped with Windows has a `curlrc` somewhere...

**Reply by ptcalex** [reply]
Thank you very much, Jay. I have resolved it: there was a `.curlrc` in the home directory with a `cacert` setting pointing to an enterprise MITM CA certificate.
